### PR TITLE
Add a code fix to add argument names

### LINKED ISSUE #9883: Request: Analyzer for named arguments
Named arguments are a great feature to write self-documented code, but if the original author didn't write them you should write every argument by hand which is kinda cumbersome. Also, when you write named arguments you might write them in a random order that makes the compiler to introduce locals to preserve side-effects in order. It would be nice to (1) have an analyzer add all named arguments for a method call (2) and suggest correction if they are not in the same order as method parameters.

### PR DESCRIPTION
**Pull request by alrz**
Fixes #9883 (previous PR: #9934)

/cc @CyrusNajmabadi

### CODE REVIEW

File: src/EditorFeatures/VisualBasicTest/CodeActions/UseNamedArguments/UseNamedArgumentsTests.vb [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+        Inherits AbstractVisualBasicCodeActionTest
+
+        Protected Overrides Function CreateCodeRefactoringProvider(workspace As Workspace) As CodeRefactoringProvider
+            Return New VisualBasicUseNamedArgumentsCodeRefactoringProvider()
+        End Function
+
+        <Fact, Trait(Traits.Feature, Traits.Features.CodeActionsUseNamedArguments)>
+        Public Async Function TestFirstArgument() As Task
+            Await TestAsync(
+NewLines("Class C \n Sub M(arg1 As Integer, arg2 As Integer) \n M([||]1, 2) \n End Sub \n End Class"),
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/VisualBasicTest/CodeActions/UseNamedArguments/UseNamedArgumentsTests.vb]: we don't do this anymore.  Please make all the tests just use normal VB multiline string literals.  Thanks!

File: src/Features/CSharp/Portable/CodeRefactorings/UseNamedArguments/CSharpUseNamedArgumentsCodeRefactoringProvider.cs [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+    [ExportCodeRefactoringProvider(LanguageNames.CSharp, Name = nameof(CSharpUseNamedArgumentsCodeRefactoringProvider)), Shared]
+    internal class CSharpUseNamedArgumentsCodeRefactoringProvider : AbstractUseNamedArgumentsCodeRefactoringProvider
+    {
+        protected override SyntaxNode GetOrSynthesizeNamedArguments(ImmutableArray<IParameterSymbol> parameters, SyntaxNode argumentList, int index)
+        {
+            switch (argumentList.Kind())
+            {
+                case SyntaxKind.ArgumentList:
+                    {
+                        var node = (ArgumentListSyntax)argumentList;
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/CodeRefactorings/UseNamedArguments/CSharpUseNamedArgumentsCodeRefactoringProvider.cs]: I feel like it must be possible to share more code between all these codepaths.

File: src/Features/CSharp/Portable/CodeRefactorings/UseNamedArguments/CSharpUseNamedArgumentsCodeRefactoringProvider.cs [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+                case SyntaxKind.BracketedArgumentList:
+                case SyntaxKind.AttributeArgumentList:
+                    return argument.Parent.Parent;
+
+                default:
+                    return null;
+            }
+        }
+
+        protected override ValueTuple<int, int> GetArgumentListIndexAndCount(SyntaxNode node)
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/CodeRefactorings/UseNamedArguments/CSharpUseNamedArgumentsCodeRefactoringProvider.cs]: you can just real tuples now :)

File: src/Features/CSharp/Portable/CodeRefactorings/UseNamedArguments/CSharpUseNamedArgumentsCodeRefactoringProvider.cs [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+                        && attributeArgument.NameEquals == null;
+
+                default:
+                    return false;
+            }
+        }
+
+        protected override bool IsLegalToAddNamedArguments(ImmutableArray<IParameterSymbol> parameters, int argumentCount)
+        {
+            return !parameters.LastOrDefault().IsParams || parameters.Length >= argumentCount;
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/CodeRefactorings/UseNamedArguments/CSharpUseNamedArgumentsCodeRefactoringProvider.cs]: LastOrDefault may return null.

File: src/Features/Core/Portable/CodeRefactorings/UseNamedArguments/AbstractUseNamedArgumentsCodeRefactoringProvider.cs [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+            }
+
+            var parameters = symbol.GetParameters();
+            if (parameters.IsDefaultOrEmpty)
+            {
+                return;
+            }
+
+            var t = GetArgumentListIndexAndCount(argument);
+            if (t.Item1 < 0)
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/CodeRefactorings/UseNamedArguments/AbstractUseNamedArgumentsCodeRefactoringProvider.cs]: with a real tuple, you can have good names here.

File: src/Features/VisualBasic/Portable/CodeRefactorings/UseNamedArguments/VisualBasicUseNamedArgumentsCodeRefactoringProvider.vb [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+            Return Not parameters.LastOrDefault().IsParams OrElse parameters.Length > argumentCount
+        End Function
+
+        Protected Overrides Function GetArgumentListIndexAndCount(node As SyntaxNode) As ValueTuple(Of Integer, Integer)
+            Dim argumentList = DirectCast(node.Parent, ArgumentListSyntax)
+            Return ValueTuple.Create(argumentList.Arguments.IndexOf(DirectCast(node, SimpleArgumentSyntax)), argumentList.Arguments.Count)
+        End Function
+
+        Protected Overrides Function GetReceiver(argument As SyntaxNode) As SyntaxNode
+            If argument.Parent?.Parent?.IsKind(SyntaxKind.Attribute) = True Then
```
Comment (CyrusNajmabadi) [inline src/Features/VisualBasic/Portable/CodeRefactorings/UseNamedArguments/VisualBasicUseNamedArgumentsCodeRefactoringProvider.vb]: argument.Parent.IsParentKind(...)

Then you don't need ?.   or   = true

File: src/Features/CSharp/Portable/CodeRefactorings/UseNamedArguments/CSharpUseNamedArgumentsCodeRefactoringProvider.cs (hunk 4 above)
Comment (alrz) [inline src/Features/CSharp/Portable/CodeRefactorings/UseNamedArguments/CSharpUseNamedArgumentsCodeRefactoringProvider.cs]: It's checked in the abstract class, should I check again?

File: src/Features/Core/Portable/CodeRefactorings/UseNamedArguments/AbstractUseNamedArgumentsCodeRefactoringProvider.cs (hunk 5 above)
Comment (alrz) [inline src/Features/Core/Portable/CodeRefactorings/UseNamedArguments/AbstractUseNamedArgumentsCodeRefactoringProvider.cs]: Does `var(index, count)` work now?

**Comment by CyrusNajmabadi** [comment]
This looks GREAT.  I definitely prefer this over the previous approach.

@Pilchie What are the options on something like this for VS17?  I'm guessing we're past the point of new features now, right?

**Comment by CyrusNajmabadi** [comment]
Note: i have some ideas on how to simplify some duplication here.  Would you mind me makign some edits?

File: src/Features/CSharp/Portable/CodeRefactorings/UseNamedArguments/CSharpUseNamedArgumentsCodeRefactoringProvider.cs (hunk 2 above)
Comment (alrz) [inline src/Features/CSharp/Portable/CodeRefactorings/UseNamedArguments/CSharpUseNamedArgumentsCodeRefactoringProvider.cs]: I felt that way myself. Just check types, there is nothing in common!

File: src/EditorFeatures/VisualBasicTest/CodeActions/UseNamedArguments/UseNamedArgumentsTests.vb (hunk 1 above)
Comment (alrz) [inline src/EditorFeatures/VisualBasicTest/CodeActions/UseNamedArguments/UseNamedArgumentsTests.vb]: Right on.

**Comment by alrz** [comment]
@CyrusNajmabadi That's ok.

**Comment by Pilchie** [comment]
It is too late for VS 2017 RTW, but a great candidate for an update.  Unfortunately, we don't have the branch for that set up quite yet, but we'll be working on doing that in the next couple of days, and it should be simple to retarget to that branch when it is up.

**Comment by alrz** [comment]
@CyrusNajmabadi I did a small tweak and clean up. Could you please take a look? Thanks.

**Comment by CyrusNajmabadi** [comment]
Your changes LGTM.

**Comment by CyrusNajmabadi** [comment]
Once we open up the post VS2017 branch, we'll merge this in.  Thanks!

**Comment by Pilchie** [comment]
I think @DustinCampbell populated `post-dev15` this morning.

**Comment by CyrusNajmabadi** [comment]
Sweet!

**Comment by DustinCampbell** [comment]
Yup. Rebase away. This is perfect!

**Comment by CyrusNajmabadi** [comment]
@DustinCampbell Tests are not running.  Is that a known issue?

**Comment by DustinCampbell** [comment]
Closed and re-opened to retrigger checks

**Comment by alrz** [comment]
tagging @CyrusNajmabadi. This is up for merge (finally!). :)

**Comment by CyrusNajmabadi** [comment]
Done!  Thanks again!

**Pull request merged**